### PR TITLE
Format code in comments following # =>

### PR DESCRIPTION
**Pull request by fschrfl**
Implements the feature discussed in  #7272.

Using the same example as in the issue:
```crystal
a = [1, 2, 3]
a.permutations    # => [[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]
a.permutations(1) # => [[1],[2],[3]]
a.permutations(2) # => [[1,2],[1,3],[2,1],[2,3],[3,1],[3,2]]
a.permutations(3) # => [[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]
a.permutations(0) # => [[]]
a.permutations(4) # => []
```
The above is now formatted like this:
```crystal
a = [1, 2, 3]
a.permutations    # => [[1, 2, 3], [1, 3, 2], [2, 1, 3], [2, 3, 1], [3, 1, 2], [3, 2, 1]]
a.permutations(1) # => [[1], [2], [3]]
a.permutations(2) # => [[1, 2], [1, 3], [2, 1], [2, 3], [3, 1], [3, 2]]
a.permutations(3) # => [[1, 2, 3], [1, 3, 2], [2, 1, 3], [2, 3, 1], [3, 1, 2], [3, 2, 1]]
a.permutations(0) # => [[]]
a.permutations(4) # => []
```

Note that the last two lines are no valid Crystal expressions in which case the formatter doesn't change them.

This is my first attempt at a contribution to the project so if I did anything wrong please let me know and I will try to improve.

### CODE REVIEW

File: src/compiler/crystal/tools/formatter.cr [hunk 1]
```diff
@@ -4516,7 +4516,15 @@ module Crystal
...
-          value = "\# => #{after_comment_value[2..-1].strip}"
+          # If a comment starts with `=>`, it indicates the result of an expression which
+          # is a valid expression itself and should be properly formatted.
+          formatted_result_expression = after_comment_value[2..-1]
+          begin
+            formatted_result_expression = Formatter.format(formatted_result_expression)
+          rescue Crystal::SyntaxException
+            # A failure to parse the comment is ignored. It is only a comment after all.
+          end
+          value = "\# => #{formatted_result_expression.strip}"
```
Comment (j8r) [inline src/compiler/crystal/tools/formatter.cr]: `#` doesn't have to be escaped.

File: src/compiler/crystal/tools/formatter.cr [hunk 2]
```diff
@@ -4516,7 +4516,15 @@ module Crystal
         raw_after_comment_value = value[1..-1]
         after_comment_value = raw_after_comment_value.strip
         if after_comment_value.starts_with?("=>")
-          value = "\# => #{after_comment_value[2..-1].strip}"
+          # If a comment starts with `=>`, it indicates the result of an expression which
+          # is a valid expression itself and should be properly formatted.
+          formatted_result_expression = after_comment_value[2..-1]
```
Comment (j8r) [inline src/compiler/crystal/tools/formatter.cr]: Can be replaced by:
`if formatted_result_expression = after_comment_value.lchop? "=>"`

Comment (j8r) [reply src/compiler/crystal/tools/formatter.cr]: To clarify, both the initial if condition and the line with `[2..-1]` can be replaced by the above expression.

**Comment by fschrfl** [comment]
I added some specs and adjusted the code from the first commit based on your suggestions. Thanks for the feedback. Hope, the specs cover everything. I am not very imaginative when it comes to test.

File: spec/compiler/formatter/formatter_spec.cr [hunk 3]
```diff
@@ -1610,4 +1610,8 @@ describe Crystal::Formatter do
       # second comment
     }
     CODE
+
+  # 7272
```
Comment (straight-shoota) [inline spec/compiler/formatter/formatter_spec.cr]: There's no need to reference the issue, a textual description is better:
```suggestion
  # format expressions in `# =>` comments
```

File: spec/compiler/formatter/formatter_spec.cr [hunk 4]
```diff
@@ -1610,4 +1610,8 @@ describe Crystal::Formatter do
...
+  assert_format "1 # => =>1+2"
```
Comment (straight-shoota) [inline spec/compiler/formatter/formatter_spec.cr]: Better use an example that's explicitly about invalid code and doesn't look like there might be some special behaviour with `=> =>` :
```suggestion
  assert_format "1 # => ?invalid  code"
```

File: src/compiler/crystal/tools/formatter.cr [hunk 5]
```diff
@@ -4515,8 +4515,15 @@ module Crystal
         value = @token.value.to_s.strip
         raw_after_comment_value = value[1..-1]
         after_comment_value = raw_after_comment_value.strip
-        if after_comment_value.starts_with?("=>")
-          value = "\# => #{after_comment_value[2..-1].strip}"
+        if formatted_result_expression = after_comment_value.lchop? "=>"
```
Comment (straight-shoota) [inline src/compiler/crystal/tools/formatter.cr]: Minor detail, but it's not formatted at the beginning, so I'd just call this variable `result_expression` (everywhere).
```suggestion
        if result_expression = after_comment_value.lchop? "=>"
```

File: spec/compiler/formatter/formatter_spec.cr (hunk 3 above)
Comment (fschrfl) [reply spec/compiler/formatter/formatter_spec.cr]: I did it this way because a comment just like that exists in several other places in the file and I wanted to be consistent. However, I do agree that a textual description is superior and I implemented that along with your other suggested changes.

File: spec/compiler/formatter/formatter_spec.cr [hunk 6]
```diff
@@ -1610,4 +1610,8 @@ describe Crystal::Formatter do
       # second comment
     }
     CODE
+
+  # format expressions in `# =>` comments
+  assert_format "1 # => ?invalid code"
+  assert_format "1 # => #=>1+2", "1 # => # => 1 + 2"
```
Comment (oprypin) [inline spec/compiler/formatter/formatter_spec.cr]: These specs cover known edge cases but it's as important to test a few basic cases first.

File: spec/compiler/formatter/formatter_spec.cr (hunk 3 above)
Comment (oprypin) [reply spec/compiler/formatter/formatter_spec.cr]: FYI commenting in that style is typical only for specs that get added specifically in response to a newly found bug.

File: spec/compiler/formatter/formatter_spec.cr (hunk 6 above)
Comment (fschrfl) [reply spec/compiler/formatter/formatter_spec.cr]: What basic cases would you suggest? The two cases that need to be covered are invalid and valid code. Invalid code is covered by the first case, for valid code I could add another basic case like `1 # => 1 + 2` but that is also already covered by the recursion in case two.

Comment (asterite) [reply spec/compiler/formatter/formatter_spec.cr]: What. happens if the comment expression isa method definition with, say, 1 + 2 as the body?

Comment (fschrfl) [reply spec/compiler/formatter/formatter_spec.cr]: That is indeed a problem I did not consider. As it is now, the formatter will put the body and `end` in separate lines which are not commented out:
```crystal
1 # => def add12() 1 + 2 end
```
turns into
```crystal
1 # => def add12
  1 + 2
end
```
It would be a good idea, anyway, to only apply formatting to certain kinds of expressions, or better still, to only apply formatting if the formatted string is equal to the result of `.inspect`ing the expression before the comment (see discussion in #7272). If that was implemented, the above problem wouldn't even come up. So before fixing this, it might be a good idea to overthink the whole approach here...

Comment (asterite) [reply spec/compiler/formatter/formatter_spec.cr]: I think it's simpler: try to format it. If you find newlines, then don't use the result.

Comment (asterite) [reply spec/compiler/formatter/formatter_spec.cr]: Of course that means that we can't always require what comes after `# => ` to be properly formatted. /cc @straight-shoota

Comment (straight-shoota) [reply spec/compiler/formatter/formatter_spec.cr]: If the formatter inserts a line break, that looks like an indicator that there is something wrong. I'm convinced it should fail in that case. I can't think of any example where the proper result of an expression described with ` =>` would be split in multiple lines.

Comment (asterite) [reply spec/compiler/formatter/formatter_spec.cr]: I don't think the code formatter should ever fail. It would be pretty annoying...

Comment (asterite) [reply spec/compiler/formatter/formatter_spec.cr]: Specially considering that it's a tool that's combined with the editor. I would find it pretty annoying if the editor starts spitting errors at me because of the way I organize my code.

Comment (straight-shoota) [reply spec/compiler/formatter/formatter_spec.cr]: Yeah maybe the formatter is not the right floor this. But it would be great to have some formal validation for `# =>` comments and similar.
Maybe that's something for ameba?

Comment (asterite) [reply spec/compiler/formatter/formatter_spec.cr]: If anything, a tool like doctest would be great.

I think having the formatter try to format the value if possible is good enough. Then doctest would actually run it and verify that the results are what you'd expect them to be.

Comment (fschrfl) [reply spec/compiler/formatter/formatter_spec.cr]: Incidentally, is there a reason why `# def add12 : Int32 1 + 2 end` and `# def add12() 1 + 2 end` are valid syntax while `# def add12 1 + 2 end` raises the syntax error "unexpected token: 1"?

Comment (asterite) [reply spec/compiler/formatter/formatter_spec.cr]: Just a lack of thought from us on those details. We should probably copy Ruby or decide what to do, but it's very low priority.
